src/System.Runtime.Extensions/src/System/Runtime/Versioning/VersioningHelper.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            ResourceScope fromResType = from & ResTypeMask;
+            ResourceScope toResType = to & ResTypeMask;
+            if (fromResType > toResType)
+                throw new ArgumentException(SR.Format(SR.Argument_ResourceScopeWrongDirection, fromResType, toResType), "from");
+
+            SxSRequirements requires = GetRequirements(to, from);
+
+            if ((requires & (SxSRequirements.AssemblyName | SxSRequirements.TypeName)) != 0 && type == null)
+                throw new ArgumentNullException("type", SR.ArgumentNull_TypeRequiredByResourceScope);
+
+            // Add in process ID, CLR base address, and appdomain ID's.  Also, use a character identifier
+            // to ensure that these can never accidentally overlap (ie, you create enough appdomains and your
+            // appdomain ID might equal your process ID).
+            StringBuilder safeName = new StringBuilder(name);
+            char separator = '_';
+            if ((requires & SxSRequirements.ProcessID) != 0) {
+                safeName.Append(separator);
+                safeName.Append('p');
+                safeName.Append(GetCurrentProcessId());
+            }
+            if ((requires & SxSRequirements.CLRInstanceID) != 0) {
+                String clrID = GetCLRInstanceString();
+                safeName.Append(separator);
+                safeName.Append('r');
+                safeName.Append(clrID);
+            }
+            // TODO -- https://github.com/dotnet/corefx/pull/12113
+            // if ((requires & SxSRequirements.AppDomainID) != 0) {
+            //     safeName.Append(separator);
+            //     safeName.Append("ad");
+            //     safeName.Append(AppDomain.CurrentDomain.Id);
+            // }
+            if ((requires & SxSRequirements.TypeName) != 0) {
+                safeName.Append(separator);
+                safeName.Append(type.Name);
+            }
+            if ((requires & SxSRequirements.AssemblyName) != 0) {
+                safeName.Append(separator);
+                safeName.Append(type.Assembly.FullName);
+            }
+            return safeName.ToString();
+        }
+
+        private static String GetCLRInstanceString()
+        {
+            // We are going to hardcode the value here to 3 (a random number) so that we don't have to 
+            // actually call GetRuntimeId() which is an ecall method and cannot be 
+            // directly called from outside of the corelib.
+            // In CoreCLR, GetRuntimeId() gets the TLS index for the thread and adds 3 to that number.
+            int id = 3;
+            return id.ToString(CultureInfo.InvariantCulture);
+        }
+
+        private static SxSRequirements GetRequirements(ResourceScope consumeAsScope, ResourceScope calleeScope)
+        {
+            SxSRequirements requires = SxSRequirements.None;
+        
+            switch(calleeScope & ResTypeMask) {
```

Comment:
Nit: The formatting of the `switch`s seems a little strange... I would expect them to look something like:

```c#
switch (calleeScope & ResTypeMask)
{
    case ResourceScope.Machine:
        switch(consumeAsScope & ResTypeMask)
        {
            case ResourceScope.Machine:
                // No work
                break;

            case ResourceScope.Process:
                requires |= SxSRequirements.ProcessID;
                break;

            ...
        }
        break;

    case ResourceScope.Process:
        ...
        break;
}
```